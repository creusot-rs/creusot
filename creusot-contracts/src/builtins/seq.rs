use creusot_contracts_proc::*;

use crate::Int;

#[creusot::builtins = "seq.Seq.seq"]
pub struct Seq<T>(std::marker::PhantomData<T>);

impl<T> Seq<T> {
    #[trusted]
    #[logic_rust]
    #[creusot::builtins = "seq.Seq.get"]
    pub fn index(self, _: Int) -> T {
        std::process::abort()
    }

    #[trusted]
    #[logic_rust]
    #[creusot::builtins = "seq.Seq.length"]
    pub fn len(self) -> Int {
        std::process::abort()
    }

    #[trusted]
    #[logic_rust]
    #[creusot::builtins = "seq.Seq.set"]
    pub fn set(self, _: Int, _: T) -> Self {
        std::process::abort()
    }

    #[trusted]
    #[logic_rust]
    #[creusot::builtins = "seq.Seq.snoc"]
    pub fn push(self, _: T) -> Self {
        std::process::abort()
    }
}
