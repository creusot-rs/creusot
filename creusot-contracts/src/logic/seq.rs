use creusot_contracts_proc::*;

use crate::logic::*;
use crate::Int;
use std::ops::Index;

#[creusot::builtins = "seq.Seq.seq"]
pub struct Seq<T: ?Sized>(std::marker::PhantomData<T>);

impl<T> EqLogic for Seq<T> {
    #[predicate]
    #[creusot::builtins = "seq.Seq.(==)"]
    fn log_eq(self, _: Self) -> bool {
        true
    }

    #[predicate]
    fn log_ne(self, o: Self) -> bool {
        pearlite! { !(self == o)}
    }

    #[logic]
    fn eq_ne(_: Self, _: Self) {}

    #[logic]
    fn refl(_: Self) {}

    #[logic]
    fn symmetry(_: Self, _: Self) {}

    #[logic]
    fn transitivity(_: Self, _: Self, _: Self) {}
}

impl<T> Seq<T> {
    #[logic]
    pub fn get(self, ix: Int) -> Option<T> {
        if ix < self.len() {
            Some(*self.index(ix))
        } else {
            None
        }
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq_ext.SeqExt.subsequence"]
    pub fn subsequence(self, _: Int, _: Int) -> Self {
        std::process::abort()
    }

    #[logic]
    pub fn tail(self) -> Self {
        self.subsequence(1, self.len())
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.length"]
    pub fn len(self) -> Int {
        std::process::abort()
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.set"]
    pub fn set(self, _: Int, _: T) -> Self {
        std::process::abort()
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.snoc"]
    pub fn push(self, _: T) -> Self {
        std::process::abort()
    }

    #[predicate]
    pub fn permutation_of(self, o: Self) -> bool {
        self.permut(o, 0, self.len())
    }

    #[trusted]
    #[predicate]
    #[creusot::builtins = "seq.Permut.permut"]
    pub fn permut(self, _: Self, _: Int, _: Int) -> bool {
        std::process::abort()
    }

    #[trusted]
    #[predicate]
    #[creusot::builtins = "seq.Permut.exchange"]
    pub fn exchange(self, _: Self, _: Int, _: Int) -> bool {
        std::process::abort()
    }
}

// A hack which allows us to use [..] notation for sequences.
// Relies on the fact we don't enforce that implementations of traits are of
// the same function type as the trait signature.. When this is addressed
// the following instance will error.
impl<T> std::ops::Index<Int> for Seq<T> {
    type Output = T;

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.get"]
    fn index(&self, _: Int) -> &T {
        std::process::abort()
    }
}
