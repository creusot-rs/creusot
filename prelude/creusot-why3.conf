[prover_modifiers]
name = "Z3"
driver = "z3_nombqi_seq.drv"