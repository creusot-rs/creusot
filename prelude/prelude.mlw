module Prelude
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double

  type usize = uint64
  type isize = int64
  type opaque_ptr

  type borrowed 'a = { current : 'a ; final : 'a; }
  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final
  val borrow_mut (a : 'a) : borrowed 'a
  ensures { *result = a }
end
