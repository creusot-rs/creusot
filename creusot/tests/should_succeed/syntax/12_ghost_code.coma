
module C12GhostCode_GhostArg
  let%span s12_ghost_code0 = "../12_ghost_code.rs" 5 28 5 44
  
  use prelude.prelude.UInt32
  
  use prelude.prelude.Int
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Intrinsic
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Snapshot
  
  let rec ghost_arg (g:Snapshot.snap_ty uint32) (return'  (ret:()))= (! bb0
    [ bb0 = s0 [ s0 =  [ &_x <- [%#s12_ghost_code0] Snapshot.new (Snapshot.inner g) ] s1 | s1 = bb1 ] 
    | bb1 = return' {_0} ]
    ) [ & _0 : () = any_l () | & g : Snapshot.snap_ty uint32 = g | & _x : Snapshot.snap_ty uint32 = any_l () ] 
    [ return' (result:())-> (! return' {result}) ]
    
end
module Core_Ptr_NonNull_NonNull_Type
  use prelude.prelude.Opaque
  
  type t_nonnull 't =
    | C_NonNull opaque_ptr
  
  function any_l (_ : 'b) : 'a
  
  let rec t_nonnull < 't > (input:t_nonnull 't) (ret  (pointer:opaque_ptr))= any
    [ good (pointer:opaque_ptr)-> {C_NonNull pointer = input} (! ret {pointer})
    | bad (pointer:opaque_ptr)-> {C_NonNull pointer <> input} {false} any ]
    
end
module Core_Marker_PhantomData_Type
  type t_phantomdata 't =
    | C_PhantomData
  
  function any_l (_ : 'b) : 'a
  
  let rec t_phantomdata < 't > (input:t_phantomdata 't) (ret  )= any
    [ good -> {C_PhantomData  = input} (! ret) | bad -> {C_PhantomData  <> input} {false} any ]
    
end
module Core_Ptr_Unique_Unique_Type
  use Core_Marker_PhantomData_Type as PhantomData'0
  
  use Core_Ptr_NonNull_NonNull_Type as NonNull'0
  
  type t_unique 't =
    | C_Unique (NonNull'0.t_nonnull 't) (PhantomData'0.t_phantomdata 't)
  
  function any_l (_ : 'b) : 'a
  
  let rec t_unique < 't > (input:t_unique 't) (ret  (pointer:NonNull'0.t_nonnull 't) (_marker:PhantomData'0.t_phantomdata 't))= any
    [ good (pointer:NonNull'0.t_nonnull 't) (_marker:PhantomData'0.t_phantomdata 't)-> {C_Unique pointer _marker
      = input}
      (! ret {pointer} {_marker})
    | bad (pointer:NonNull'0.t_nonnull 't) (_marker:PhantomData'0.t_phantomdata 't)-> {C_Unique pointer _marker
      <> input}
      {false}
      any ]
    
end
module Alloc_RawVec_Cap_Type
  use prelude.prelude.UIntSize
  
  use prelude.prelude.Int
  
  type t_cap  =
    | C_Cap usize
  
  function any_l (_ : 'b) : 'a
  
  let rec t_cap (input:t_cap) (ret  (field_0:usize))= any
    [ good (field_0:usize)-> {C_Cap field_0 = input} (! ret {field_0})
    | bad (field_0:usize)-> {C_Cap field_0 <> input} {false} any ]
    
end
module Alloc_RawVec_RawVec_Type
  use Alloc_RawVec_Cap_Type as Cap'0
  
  use Core_Ptr_Unique_Unique_Type as Unique'0
  
  type t_rawvec 't 'a =
    | C_RawVec (Unique'0.t_unique 't) (Cap'0.t_cap) 'a
  
  function any_l (_ : 'b) : 'a
  
  let rec t_rawvec < 't > < 'a > (input:t_rawvec 't 'a) (ret  (ptr:Unique'0.t_unique 't) (cap:Cap'0.t_cap) (alloc:'a))= any
    [ good (ptr:Unique'0.t_unique 't) (cap:Cap'0.t_cap) (alloc:'a)-> {C_RawVec ptr cap alloc = input}
      (! ret {ptr} {cap} {alloc})
    | bad (ptr:Unique'0.t_unique 't) (cap:Cap'0.t_cap) (alloc:'a)-> {C_RawVec ptr cap alloc <> input} {false} any ]
    
end
module Alloc_Vec_Vec_Type
  use prelude.prelude.UIntSize
  
  use prelude.prelude.Int
  
  use Alloc_RawVec_RawVec_Type as RawVec'0
  
  type t_vec 't 'a =
    | C_Vec (RawVec'0.t_rawvec 't 'a) usize
  
  function any_l (_ : 'b) : 'a
  
  let rec t_vec < 't > < 'a > (input:t_vec 't 'a) (ret  (buf:RawVec'0.t_rawvec 't 'a) (len:usize))= any
    [ good (buf:RawVec'0.t_rawvec 't 'a) (len:usize)-> {C_Vec buf len = input} (! ret {buf} {len})
    | bad (buf:RawVec'0.t_rawvec 't 'a) (len:usize)-> {C_Vec buf len <> input} {false} any ]
    
end
module Alloc_Alloc_Global_Type
  type t_global  =
    | C_Global
  
  function any_l (_ : 'b) : 'a
  
  let rec t_global (input:t_global) (ret  )= any
    [ good -> {C_Global  = input} (! ret) | bad -> {C_Global  <> input} {false} any ]
    
end
module C12GhostCode_GhostVec
  let%span s12_ghost_code0 = "../12_ghost_code.rs" 9 22 9 32
  
  let%span s12_ghost_code1 = "../12_ghost_code.rs" 10 35 10 50
  
  let%span span2 = "../../../../../creusot-contracts/src/invariant.rs" 8 8 8 12
  
  let%span span3 = "" 0 0 0 0
  
  let%span span4 = "../../../../../creusot-contracts/src/std/vec.rs" 19 21 19 25
  
  let%span span5 = "../../../../../creusot-contracts/src/std/vec.rs" 18 14 18 41
  
  let%span span6 = "../../../../../creusot-contracts/src/std/vec.rs" 19 4 19 36
  
  let%span span7 = "../../../../../creusot-contracts/src/std/vec.rs" 60 20 60 41
  
  let%span span8 = "../../../../../creusot-contracts/src/resolve.rs" 46 8 46 12
  
  let%span span9 = "../../../../../creusot-contracts/src/logic/ops.rs" 20 8 20 31
  
  let%span span10 = "../../../../../creusot-contracts/src/std/vec.rs" 51 8 51 85
  
  let%span span11 = "../../../../../creusot-contracts/src/std/vec.rs" 69 26 69 44
  
  let%span span12 = "" 0 0 0 0
  
  use prelude.prelude.UInt32
  
  use seq.Seq
  
  predicate invariant'1 (self : Seq.seq uint32) =
    [%#span2] true
  
  predicate inv'1 (_x : Seq.seq uint32)
  
  axiom inv'1 : forall x : Seq.seq uint32 . inv'1 x = true
  
  use Alloc_Alloc_Global_Type as Global'0
  
  use Alloc_Vec_Vec_Type as Vec'0
  
  use prelude.prelude.UIntSize
  
  use prelude.prelude.UIntSize
  
  use prelude.prelude.Int
  
  constant max'0 : usize = [%#span3] (18446744073709551615 : usize)
  
  use seq.Seq
  
  predicate inv'0 (_x : Vec'0.t_vec uint32 (Global'0.t_global))
  
  function shallow_model'0 (self : Vec'0.t_vec uint32 (Global'0.t_global)) : Seq.seq uint32
  
  axiom shallow_model'0_spec : forall self : Vec'0.t_vec uint32 (Global'0.t_global) . ([%#span4] inv'0 self)
   -> ([%#span6] inv'1 (shallow_model'0 self))
  && ([%#span5] Seq.length (shallow_model'0 self) <= UIntSize.to_int (max'0 : usize))
  
  predicate invariant'0 (self : Vec'0.t_vec uint32 (Global'0.t_global)) =
    [%#span7] inv'1 (shallow_model'0 self)
  
  axiom inv'0 : forall x : Vec'0.t_vec uint32 (Global'0.t_global) . inv'0 x = true
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Intrinsic
  
  use prelude.prelude.Snapshot
  
  predicate resolve'1 (self : uint32) =
    [%#span8] true
  
  use seq.Seq
  
  function index_logic'0 [@inline:trivial] (self : Vec'0.t_vec uint32 (Global'0.t_global)) (ix : int) : uint32 =
    [%#span9] Seq.get (shallow_model'0 self) ix
  
  predicate resolve'0 (self : Vec'0.t_vec uint32 (Global'0.t_global)) =
    [%#span10] forall i : int . 0 <= i /\ i < Seq.length (shallow_model'0 self)  -> resolve'1 (index_logic'0 self i)
  
  let rec new'0 (_1:()) (return'  (ret:Vec'0.t_vec uint32 (Global'0.t_global)))= any
    [ return' (result:Vec'0.t_vec uint32 (Global'0.t_global))-> {[%#span12] inv'0 result}
      {[%#span11] Seq.length (shallow_model'0 result) = 0}
      (! return' {result}) ]
    
  
  let rec ghost_vec (_1:()) (return'  (ret:()))= (! bb0
    [ bb0 = s0
      [ s0 = new'0 {[%#s12_ghost_code0] ()} (fun (_ret':Vec'0.t_vec uint32 (Global'0.t_global)) ->  [ &x <- _ret' ] s1)
      | s1 = bb1 ]
      
    | bb1 = s0 [ s0 = -{resolve'0 x}- s1 | s1 =  [ &_s <- [%#s12_ghost_code1] Snapshot.new x ] s2 | s2 = bb2 ] 
    | bb2 = bb3
    | bb3 = return' {_0} ]
    )
    [ & _0 : () = any_l ()
    | & x : Vec'0.t_vec uint32 (Global'0.t_global) = any_l ()
    | & _s : Snapshot.snap_ty (Vec'0.t_vec uint32 (Global'0.t_global)) = any_l () ]
     [ return' (result:())-> (! return' {result}) ] 
end
module C12GhostCode_GhostCopy
  let%span s12_ghost_code0 = "../12_ghost_code.rs" 18 12 18 13
  
  let%span s12_ghost_code1 = "../12_ghost_code.rs" 19 17 19 52
  
  let%span s12_ghost_code2 = "../12_ghost_code.rs" 20 9 20 33
  
  use prelude.prelude.Intrinsic
  
  use prelude.prelude.Int32
  
  use seq.Seq
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Snapshot
  
  use seq.Seq
  
  use seq.Seq
  
  use prelude.prelude.Int
  
  let rec ghost_copy (_1:()) (return'  (ret:()))= (! bb0
    [ bb0 = s0
      [ s0 =  [ &a <- [%#s12_ghost_code0] (0 : int32) ] s1
      | s1 =  [ &_s <- [%#s12_ghost_code1] Snapshot.new (Seq.snoc (Seq.empty  : Seq.seq int32) (0 : int32)) ] s2
      | s2 = bb1 ]
      
    | bb1 = s0 [ s0 =  [ &_4 <- [%#s12_ghost_code2] Snapshot.new (Seq.snoc (Snapshot.inner _s) a) ] s1 | s1 = bb2 ] 
    | bb2 = s0
      [ s0 =  [ &_s <- _4 ] (any [ any_ (_any:Snapshot.snap_ty (Seq.seq int32))-> (!  [ &_4 <- _any ] s1) ] )
      | s1 = return' {_0} ]
       ]
    )
    [ & _0 : () = any_l ()
    | & a : int32 = any_l ()
    | & _s : Snapshot.snap_ty (Seq.seq int32) = any_l ()
    | & _4 : Snapshot.snap_ty (Seq.seq int32) = any_l () ]
     [ return' (result:())-> (! return' {result}) ] 
end
module C12GhostCode_GhostIsCopy
  let%span s12_ghost_code0 = "../12_ghost_code.rs" 24 16 24 17
  
  let%span s12_ghost_code1 = "../12_ghost_code.rs" 26 12 26 27
  
  let%span s12_ghost_code2 = "../12_ghost_code.rs" 29 18 29 26
  
  let%span span3 = "../../../../../creusot-contracts/src/resolve.rs" 26 20 26 34
  
  use prelude.prelude.Int32
  
  use prelude.prelude.Borrow
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Intrinsic
  
  use prelude.prelude.Snapshot
  
  predicate resolve'0 (self : borrowed int32) =
    [%#span3] self.final = self.current
  
  use prelude.prelude.Int
  
  let rec ghost_is_copy (_1:()) (return'  (ret:()))= (! bb0
    [ bb0 = s0
      [ s0 =  [ &x <- [%#s12_ghost_code0] (0 : int32) ] s1
      | s1 = Borrow.borrow_mut <int32> {x} (fun (_ret':borrowed int32) ->  [ &r <- _ret' ]  [ &x <- r.final ] s2)
      | s2 = -{resolve'0 r}- s3
      | s3 =  [ &g <- [%#s12_ghost_code1] Snapshot.new r ] s4
      | s4 = bb1 ]
      
    | bb1 = s0
      [ s0 =  [ &g1 <- g ] s1
      | s1 =  [ &g2 <- g ] s2
      | s2 = {[@expl:assertion] [%#s12_ghost_code2] g1 = g2} s3
      | s3 = return' {_0} ]
       ]
    )
    [ & _0 : () = any_l ()
    | & x : int32 = any_l ()
    | & r : borrowed int32 = any_l ()
    | & g : Snapshot.snap_ty (borrowed int32) = any_l ()
    | & g1 : Snapshot.snap_ty (borrowed int32) = any_l ()
    | & g2 : Snapshot.snap_ty (borrowed int32) = any_l () ]
     [ return' (result:())-> (! return' {result}) ] 
end
module C12GhostCode_GhostCheck
  let%span s12_ghost_code0 = "../12_ghost_code.rs" 36 16 36 26
  
  let%span s12_ghost_code1 = "../12_ghost_code.rs" 39 4 39 31
  
  let%span s12_ghost_code2 = "../12_ghost_code.rs" 41 11 41 12
  
  let%span s12_ghost_code3 = "../12_ghost_code.rs" 43 23 43 24
  
  let%span s12_ghost_code4 = "../12_ghost_code.rs" 43 4 43 25
  
  let%span span5 = "../../../../../creusot-contracts/src/invariant.rs" 8 8 8 12
  
  let%span span6 = "" 0 0 0 0
  
  let%span span7 = "../../../../../creusot-contracts/src/std/vec.rs" 19 21 19 25
  
  let%span span8 = "../../../../../creusot-contracts/src/std/vec.rs" 18 14 18 41
  
  let%span span9 = "../../../../../creusot-contracts/src/std/vec.rs" 19 4 19 36
  
  let%span span10 = "../../../../../creusot-contracts/src/std/vec.rs" 60 20 60 41
  
  let%span span11 = "../../../../../creusot-contracts/src/resolve.rs" 46 8 46 12
  
  let%span span12 = "../../../../../creusot-contracts/src/logic/ops.rs" 20 8 20 31
  
  let%span span13 = "../../../../../creusot-contracts/src/std/vec.rs" 51 8 51 85
  
  let%span span14 = "../../../../../creusot-contracts/src/model.rs" 90 8 90 31
  
  let%span span15 = "" 0 0 0 0
  
  let%span span16 = "../../../../../creusot-contracts/src/std/vec.rs" 78 26 78 48
  
  let%span span17 = "../../../../../creusot-contracts/src/model.rs" 108 8 108 31
  
  let%span span18 = "" 0 0 0 0
  
  let%span span19 = "" 0 0 0 0
  
  let%span span20 = "../../../../../creusot-contracts/src/std/vec.rs" 82 26 82 51
  
  let%span span21 = "../12_ghost_code.rs" 32 0 32 8
  
  let%span span22 = "../../../../../creusot-contracts/src/std/vec.rs" 69 26 69 44
  
  let%span span23 = "" 0 0 0 0
  
  use prelude.prelude.Int32
  
  use seq.Seq
  
  predicate invariant'4 (self : Seq.seq int32) =
    [%#span5] true
  
  predicate inv'4 (_x : Seq.seq int32)
  
  axiom inv'4 : forall x : Seq.seq int32 . inv'4 x = true
  
  use Alloc_Alloc_Global_Type as Global'0
  
  use Alloc_Vec_Vec_Type as Vec'0
  
  predicate invariant'3 (self : Vec'0.t_vec int32 (Global'0.t_global)) =
    [%#span5] true
  
  predicate inv'3 (_x : Vec'0.t_vec int32 (Global'0.t_global))
  
  axiom inv'3 : forall x : Vec'0.t_vec int32 (Global'0.t_global) . inv'3 x = true
  
  predicate invariant'2 (self : int32) =
    [%#span5] true
  
  predicate inv'2 (_x : int32)
  
  axiom inv'2 : forall x : int32 . inv'2 x = true
  
  use prelude.prelude.Borrow
  
  predicate invariant'1 (self : borrowed (Vec'0.t_vec int32 (Global'0.t_global))) =
    [%#span5] true
  
  predicate inv'1 (_x : borrowed (Vec'0.t_vec int32 (Global'0.t_global)))
  
  axiom inv'1 : forall x : borrowed (Vec'0.t_vec int32 (Global'0.t_global)) . inv'1 x = true
  
  use prelude.prelude.UIntSize
  
  use prelude.prelude.UIntSize
  
  use prelude.prelude.Int
  
  constant max'0 : usize = [%#span6] (18446744073709551615 : usize)
  
  use seq.Seq
  
  predicate inv'0 (_x : Vec'0.t_vec int32 (Global'0.t_global))
  
  function shallow_model'0 (self : Vec'0.t_vec int32 (Global'0.t_global)) : Seq.seq int32
  
  axiom shallow_model'0_spec : forall self : Vec'0.t_vec int32 (Global'0.t_global) . ([%#span7] inv'0 self)
   -> ([%#span9] inv'4 (shallow_model'0 self))
  && ([%#span8] Seq.length (shallow_model'0 self) <= UIntSize.to_int (max'0 : usize))
  
  predicate invariant'0 (self : Vec'0.t_vec int32 (Global'0.t_global)) =
    [%#span10] inv'4 (shallow_model'0 self)
  
  axiom inv'0 : forall x : Vec'0.t_vec int32 (Global'0.t_global) . inv'0 x = true
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Intrinsic
  
  predicate resolve'1 (self : int32) =
    [%#span11] true
  
  use seq.Seq
  
  function index_logic'0 [@inline:trivial] (self : Vec'0.t_vec int32 (Global'0.t_global)) (ix : int) : int32 =
    [%#span12] Seq.get (shallow_model'0 self) ix
  
  predicate resolve'0 (self : Vec'0.t_vec int32 (Global'0.t_global)) =
    [%#span13] forall i : int . 0 <= i /\ i < Seq.length (shallow_model'0 self)  -> resolve'1 (index_logic'0 self i)
  
  function shallow_model'2 (self : Vec'0.t_vec int32 (Global'0.t_global)) : Seq.seq int32 =
    [%#span14] shallow_model'0 self
  
  let rec len'0 (self:Vec'0.t_vec int32 (Global'0.t_global)) (return'  (ret:usize))= {[@expl:precondition] [%#span15] inv'3 self}
    any
    [ return' (result:usize)-> {[%#span16] UIntSize.to_int result = Seq.length (shallow_model'2 self)}
      (! return' {result}) ]
    
  
  use seq.Seq
  
  function shallow_model'1 (self : borrowed (Vec'0.t_vec int32 (Global'0.t_global))) : Seq.seq int32 =
    [%#span17] shallow_model'0 self.current
  
  let rec push'0 (self:borrowed (Vec'0.t_vec int32 (Global'0.t_global))) (value:int32) (return'  (ret:()))= {[@expl:precondition] [%#span19] inv'2 value}
    {[@expl:precondition] [%#span18] inv'1 self}
    any
    [ return' (result:())-> {[%#span20] shallow_model'0 self.final = Seq.snoc (shallow_model'1 self) value}
      (! return' {result}) ]
    
  
  use prelude.prelude.Snapshot
  
  function logi_drop'0 [#"../12_ghost_code.rs" 33 0 33 21] (_1 : Vec'0.t_vec int32 (Global'0.t_global)) : () =
    [%#span21] ()
  
  let rec new'0 (_1:()) (return'  (ret:Vec'0.t_vec int32 (Global'0.t_global)))= any
    [ return' (result:Vec'0.t_vec int32 (Global'0.t_global))-> {[%#span23] inv'0 result}
      {[%#span22] Seq.length (shallow_model'0 result) = 0}
      (! return' {result}) ]
    
  
  let rec ghost_check (_1:()) (return'  (ret:()))= (! bb0
    [ bb0 = s0
      [ s0 = new'0 {[%#s12_ghost_code0] ()} (fun (_ret':Vec'0.t_vec int32 (Global'0.t_global)) ->  [ &x <- _ret' ] s1)
      | s1 = bb1 ]
      
    | bb1 = s0 [ s0 =  [ &_2 <- [%#s12_ghost_code1] Snapshot.new (let _ = logi_drop'0 x in ()) ] s1 | s1 = bb2 ] 
    | bb2 = s0
      [ s0 = Borrow.borrow_mut <Vec'0.t_vec int32 (Global'0.t_global)> {x}
          (fun (_ret':borrowed (Vec'0.t_vec int32 (Global'0.t_global))) ->  [ &_5 <- _ret' ]  [ &x <- _5.final ] s1)
      | s1 = push'0 {_5} {[%#s12_ghost_code2] (0 : int32)} (fun (_ret':()) ->  [ &_4 <- _ret' ] s2)
      | s2 = bb3 ]
      
    | bb3 = s0 [ s0 = len'0 {x} (fun (_ret':usize) ->  [ &_8 <- _ret' ] s1) | s1 = bb4 ] 
    | bb4 = s0
      [ s0 = -{resolve'0 x}- s1
      | s1 = UIntSize.eq {_8} {[%#s12_ghost_code3] (1 : usize)}
          (fun (_ret':bool) ->  [ &_7 <- _ret' ] (any [ any_ (_any:usize)-> (!  [ &_8 <- _any ] s2) ] ))
      | s2 = any [ br0 -> {_7 = false} (! bb6) | br1 -> {_7} (! bb5) ]  ]
      
    | bb5 = bb7
    | bb7 = return' {_0}
    | bb6 = {[%#s12_ghost_code4] false} any ]
    )
    [ & _0 : () = any_l ()
    | & x : Vec'0.t_vec int32 (Global'0.t_global) = any_l ()
    | & _2 : Snapshot.snap_ty () = any_l ()
    | & _4 : () = any_l ()
    | & _5 : borrowed (Vec'0.t_vec int32 (Global'0.t_global)) = any_l ()
    | & _7 : bool = any_l ()
    | & _8 : usize = any_l () ]
     [ return' (result:())-> (! return' {result}) ] 
end
module C12GhostCode_MyStruct_Type
  use prelude.prelude.UInt32
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Int
  
  type t_mystruct  =
    | C_MyStruct uint32 (Snapshot.snap_ty uint32)
  
  function any_l (_ : 'b) : 'a
  
  let rec t_mystruct (input:t_mystruct) (ret  (f:uint32) (g:Snapshot.snap_ty uint32))= any
    [ good (f:uint32) (g:Snapshot.snap_ty uint32)-> {C_MyStruct f g = input} (! ret {f} {g})
    | bad (f:uint32) (g:Snapshot.snap_ty uint32)-> {C_MyStruct f g <> input} {false} any ]
    
  
  function mystruct_f (self : t_mystruct) : uint32 =
    match self with
      | C_MyStruct a _ -> a
      end
  
  function mystruct_g (self : t_mystruct) : Snapshot.snap_ty uint32 =
    match self with
      | C_MyStruct _ a -> a
      end
end
module C12GhostCode_TakesStruct
  let%span s12_ghost_code0 = "../12_ghost_code.rs" 53 10 53 27
  
  let%span s12_ghost_code1 = "../12_ghost_code.rs" 51 11 51 20
  
  let%span span2 = "../../../../../creusot-contracts/src/model.rs" 90 8 90 31
  
  let%span span3 = "../../../../../creusot-contracts/src/snapshot.rs" 28 20 28 48
  
  use prelude.prelude.Borrow
  
  use prelude.prelude.UInt32
  
  use prelude.prelude.Int
  
  use prelude.prelude.UInt32
  
  function shallow_model'1 (self : uint32) : int =
    [%#span2] UInt32.to_int self
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Snapshot
  
  function shallow_model'0 (self : Snapshot.snap_ty uint32) : int =
    [%#span3] shallow_model'1 (Snapshot.inner self)
  
  use C12GhostCode_MyStruct_Type as C12GhostCode_MyStruct_Type
  
  use prelude.prelude.Intrinsic
  
  use prelude.prelude.Int
  
  use C12GhostCode_MyStruct_Type as MyStruct'0
  
  use prelude.prelude.Snapshot
  
  let rec takes_struct (x:MyStruct'0.t_mystruct) (return'  (ret:()))= {[%#s12_ghost_code1] shallow_model'0 (C12GhostCode_MyStruct_Type.mystruct_g x)
    = 0}
    (! bb0
    [ bb0 = s0
      [ s0 =  [ &_3 <- [%#s12_ghost_code0] Snapshot.new (C12GhostCode_MyStruct_Type.mystruct_f x) ] s1 | s1 = bb1 ]
      
    | bb1 = s0
      [ s0 = MyStruct'0.t_mystruct {x}
          (fun (l_f'0:uint32) (l_g'0:Snapshot.snap_ty uint32) ->
             [ &x <- MyStruct'0.C_MyStruct l_f'0 _3 ] 
            (any [ any_ (_any:Snapshot.snap_ty uint32)-> (!  [ &_3 <- _any ] s1) ] ))
      | s1 = return' {_0} ]
       ]
    ) [ & _0 : () = any_l () | & x : MyStruct'0.t_mystruct = x | & _3 : Snapshot.snap_ty uint32 = any_l () ] 
    [ return' (result:())-> (! return' {result}) ]
    
end
