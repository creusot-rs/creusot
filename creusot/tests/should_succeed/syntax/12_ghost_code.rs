extern crate creusot_contracts;
use creusot_contracts::{logic::Seq, *};

pub fn ghost_arg(g: Ghost<u32>) {
    let _x: Ghost<u32> = ghost! { *g };
}

pub fn ghost_vec() {
    let x: Vec<u32> = Vec::new();
    let mut _s: Ghost<Vec<_>> = ghost! { x };
}

#[open]
#[logic]
pub fn omg() {}

pub fn ghost_copy() {
    let a = 0;
    let mut _s = ghost! { Seq::EMPTY.push(0) };
    _s = ghost! { { _s.push(a) } };
}

#[logic]
fn logi_drop<T>(_: T) {}

pub fn ghost_check() {
    let mut x = Vec::new();

    // We ghost capture the value and then drop it without affecting program
    ghost! { {logi_drop(x); } };

    x.push(0);

    assert!(x.len() == 1);
}

pub struct MyStruct {
    f: u32,
    g: Ghost<u32>,
}

#[requires(x.g@ == 0)]
pub fn takes_struct(mut x: MyStruct) {
    x.g = ghost! { x.f };
}
