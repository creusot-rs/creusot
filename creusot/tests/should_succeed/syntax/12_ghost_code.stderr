warning: unnecessary braces around function argument
  --> 12_ghost_code.rs:20:19
   |
20 |     _s = ghost! { { _s.push(a) } };
   |                   ^^          ^^
   |
   = note: `#[warn(unused_braces)]` on by default
help: remove these braces
   |
20 -     _s = ghost! { { _s.push(a) } };
20 +     _s = ghost! { _s.push(a) };
   |

warning: 1 warning emitted

